### ISSUE TITLE
Mouse Track Randomizer?

### ISSUE THREAD

**Issue by 20minsmailacc1012545645** [mrmin123/kancolle-auto#52]
Does the tool have such thing built-in like some other japanese or chinese tools?

**Comment by mrmin123** [comment]
I'm not familiar with the other Japanese/Chinese tools; could you expand on what they do for their 'mouse track randomizer'? Random mouse movement? Random click locations?

This tool has two randomization elements implemented:

1. When refreshing the home page, it will use a random page to do it, meaning that it will access either the user profile, the quests screen, sortie screen, repair screen, or resupply screen, and then go back to the home page. (More pages can be supported, I just haven't gotten around to supporting them)
2. ALL sleep timers (basically, whenever it waits before it presses the next button) are actually waits for a random value between a set range. Wherever you see the `sleep(n)` command in the code, the sleep length is a random value between n and n*2. Wherever you see the `get_rand(n, m)` command in the code, the sleep length is a random value between n and n + m.

I didn't implement randomness in the mouse movement and clicks because it seemed ridiculous that DMM or whoever would track that information (that's a lot of extra bandwidth for them!), but that's pure conjecture. If there's enough support for it and/or evidence that they track mouse movement/mouse click locations I'll move to implement something for it.

**Comment by perryhuynh** [comment]
Random click locations would be nice to have, you can probably reference the code from Onegai-ooyodosan here: https://github.com/tantinevincent/Onegai-ooyodosan/blob/master/common.sikuli/common.py if you ever wanted to put it in

**Comment by kevin01523** [comment]
its not hard to implement but sometimes it clicks on wrong buttons i even need to edit some parts for me to use it like correcting some offsets and updating images

**Comment by 20minsmailacc1012545645** [comment]
Both random movement and random click location. On Chinese forum there were a report that after an update few months ago (If I remembered correctly, that's back in May), dev can now track mouse pointer movement.

**Comment by 20minsmailacc1012545645** [comment]
Edit: I just googled it, and found a post mentioning about this: http://kancollebot.pixnet.net/blog/post/288508168 (in Chinese). It said after May update, if you click several buttons the client would send the mouse pointer x and y location info to server. However there are disagreement on what button click would trigger sending location, some said it is when you click the large sortie or expedition button, some said when you confirm it, and etc. The link also include some links to Japanese/Chinese tools that you may want to take a look at.

Also, if only that two place have included randomness, then I think there should be a few other place that also include a bit randomness.For instance, delay between every actions should have a minor fluctuations instead of constant or zero.

**Comment by mrmin123** [comment]
I could barely make out what was being talked about in that link via Google Translate, but I'll look into random click locations, but probably not the mouse movement. Seems a bit overkill at the moment.

And for delay between every action, the`sleep()` method I mentioned previously (the one with the random timer) is called between practically every action, so that's in there already.

**Comment by Laforeta** [comment]
They have been tracking mouse clicks since June, you can see this yourself by decompiling SallyMain.swf and go to scene.sally.views.mission.MissionMainView 

      import common.util.Util;
    ......
          private function _handleClickMissionItem(param1:MissionItem, param2:int) : void
      {
         Util.$$(this);
         _selectedMissionItem = param1;
         _setEnseiRightView(param1.missionData,param2);
      }
    ......

If one track this class all the way back to core.swf, one will be able to see that it takes the mouse location and time when the mouse event is fired during expedition selection and encode it before sending it to the server in the api_serial__id string. 

I doubt this parameter is used for screening or even collected by default, but for an account already flagged for suspected botting it is trivial to collect this data for a fixed certain time period, analyse the data and find out that this user is magically clicking the same spot all the time.

**Comment by mrmin123** [comment]
Thanks for the confirmation @Laforeta. Sounds like I should implement the randomizing the mouse click locations sooner, rather than later.

**Comment by mrmin123** [comment]
Anybody who is interested in the random mouse click location functionality, please see this branch: [52_mouse_randomizer](https://github.com/mrmin123/kancolle-auto/tree/52_mouse_randomizer)

It selects a random location based on the image/matching pattern, so it should theoretically never press the wrong button. Many of the images have been updated to account for this, but might still have some deadzones. If you find any, please let me know.

**Comment by kevin01523** [comment]
testing it now while playing some vn

**Comment by lolmon50** [comment]
So, I have been running a fork of one of your Septermberish commits with various changes (one being far larger random delays) as that was my primary theory for what their bot checker looks for.  In retrospect, I should do some more in depth analysis of the game based on @Laforeta post, as I only ever saw the actual requests sent to the server, and never once thought that might be in the api_serial__id.  Either way, that account is now banned, so it was probably a good addition to have.

**Comment by mrmin123** [comment]
Ouch, sorry for your loss @lolmon50. Two quick questions: Did you bot prior to using a fork of this project? Did you leave the bot running close to 24/7?

Either way, I'm going to try and push this into the main branch sometime tonight...

**Comment by lolmon50** [comment]
I did not, pretty much an 8-5 operation, your fork was the first framework I found that seemed modern enough to actually bother tinkering with.  I will be interested to hear if I am the only one though.

The only reason why I started botting was that I had done everything else in the game that interested me, and grinding between events was no longer fun.  So Its not a big loss, but I might make a new proper "test" account to continue development.

**Comment by mrmin123** [comment]
I've merged the random mouse click location functionality into the main branch as of f5add82753e4acd65854412f31ecb7f3f5b2dcb0, but will keep this thread open to discuss further anti-detection enhancements/ideas.

**Comment by kevin01523** [comment]
before using this one I'm only using the exped one and still fine and it has no randomization whatsoever

**Comment by kevin01523** [comment]
i noticed sometimes the randomized clicking prevents the mouse to click the fleet number when sending expeditions and i guess adding more wait time would fix many problems when using a slow pc

**Comment by mrmin123** [comment]
@kevin01523 I did tighten up the fleet number images a few days ago; could you check to see if you're up to date? If you are, I'll take another look at them tonight.

**Comment by kevin01523** [comment]
yes im up to date im using the direct clone from your rep so im sure well it always happen on fleet 4

**Comment by mrmin123** [comment]
The latest merge to master (d8e2139) today (2015/10/25) implements random walking through the menus for further bot-detection evasion, among other improvements and bugfixes.

Basically, the new `rnvaigation` method allows the script to go to its next destination screen (sortie screen, resupply screen, repair screen, etc) either directly or after visiting 1-3 random screens before it. This hopefully better mimics the player who has their brain turned off while playing.

Please see the [changelog](https://github.com/mrmin123/kancolle-auto/wiki/Changelog#2015-10-25) for the full list of changes.

**Comment by mrmin123** [comment]
Expanded click range to cover more of screen when pressing `compass.png`, `next.png`, and `next_alt.png` in e065a00.
